src/Workspaces/Core/Portable/Utilities/ReferenceHolder`1.cs

```diff
@@ -0,0 +1,111 @@
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Roslyn.Utilities
+{
+    internal readonly struct ReferenceHolder<T> : IEquatable<ReferenceHolder<T>>
+        where T : class?
+    {
+        [AllowNull, MaybeNull]
+        private readonly T _strongReference;
+        private readonly WeakReference<T>? _weakReference;
+        private readonly int _hashCode;
+
+        private ReferenceHolder(T strongReference)
+        {
+            _strongReference = strongReference;
+            _weakReference = null;
+            _hashCode = 0;
+        }
+
+        private ReferenceHolder(WeakReference<T> weakReference, int hashCode)
+        {
+            _strongReference = null;
+            _weakReference = weakReference;
+            _hashCode = hashCode;
+        }
+
+        public static ReferenceHolder<T> Strong(T value)
+            => new ReferenceHolder<T>(value);
+
+        public static ReferenceHolder<T> Weak(T value)
+        {
+            if (value is null)
+            {
+                // Track this as a strong reference so we know 'Equals' should only look at values originally null.
+                return Strong(value);
+            }
+
+            return new ReferenceHolder<T>(new WeakReference<T>(value), EqualityComparer<T>.Default.GetHashCode(value));
+        }
+
+        [return: MaybeNull]
+        public T TryGetTarget()
+        {
+            if (_weakReference is object)
+                return _weakReference.GetTarget();
+
+            return _strongReference;
+        }
+
+        public override bool Equals(object? obj)
+        {
+            return obj is ReferenceHolder<T> other
+                && Equals(other);
+        }
+
+        public bool Equals(ReferenceHolder<T> other)
+        {
+            var x = TryGetTarget();
+            var y = other.TryGetTarget();
+            if (x is null)
+            {
+                // Null values are equal iff both were originally references to null.
+                return y is null && other._weakReference is null;
+            }
+
+            // Intentional reference equality check
+            return x == y;
+        }
+
+        public override int GetHashCode()
+        {
+            if (_weakReference is object)
+                return _hashCode;
+
+            return _strongReference?.GetHashCode() ?? 0;
+        }
+
+        internal TestAccessor GetTestAccessor()
+        {
+            return new TestAccessor(this);
+        }
+
+        internal readonly struct TestAccessor
+        {
+#pragma warning disable IDE0052 // Remove unread private members
+            private readonly ReferenceHolder<T> _referenceHolder;
+#pragma warning restore IDE0052 // Remove unread private members
+
+            internal TestAccessor(ReferenceHolder<T> referenceHolder)
+            {
+                _referenceHolder = referenceHolder;
+            }
```

Comment:
Why have instance test accessors that then aren't used?

Reply:
It's the common pattern generated by the refactorings. We just need an IDE0052 suppressor to match this.

Reply:
I'm not sure why we're using a refactoring to generate code we're not using...

Reply:
The refactoring generated the test accessor, which I then added a static member to.

Reply:
I guess I don't care if the refactoring is generating code we're not using...we should delete it.

Reply:
➡️ Removed this